src/channel_internal.c

```diff
@@ -74,31 +74,48 @@ static void channel_internal_dispose(CHANNEL_INTERNAL* channel_internal)
     sm_destroy(channel_internal->sm);
 }
 
+
+static int copy_entry(PDLIST_ENTRY listEntry, void* actionContext, bool* continueProcessing)
+{
+    DLIST_ENTRY* op_list = (DLIST_ENTRY*)actionContext;
+    DList_InsertTailList(op_list, listEntry);
+    *continueProcessing = true;
+    return 0;
+}
+
 static void abandon_pending_operations(void* context)
 {
     CHANNEL_INTERNAL* channel_internal_ptr = context;
 
+    DLIST_ENTRY op_list;
+    DList_InitializeListHead(&op_list);
     /*Codes_SRS_CHANNEL_INTERNAL_43_167: [abandon_pending_operations shall call srw_lock_acquire_exclusive.]*/
     srw_lock_acquire_exclusive(channel_internal_ptr->lock);
     {
         /*Codes_SRS_CHANNEL_INTERNAL_43_168: [abandon_pending_operations shall set is_open to false.]*/
         channel_internal_ptr->is_open = false;
 
+        /*Codes_SRS_CHANNEL_INTERNAL_43_174: [abandon_pending_operations shall make a local copy of the list of pending operations.]*/
+        (void)DList_ForEach(&channel_internal_ptr->op_list, copy_entry, &op_list);
```

Comment:
>DList_ForEach [](http://example.com/codeflow?start=14&length=13)

This looks like it iteratively adding all of the items to the new list without properly removing them from the source list. 

It looks like this corrupting the source list as it copies them into the new list. 

It's assuming that the iteration over the source list will continue to work as you're destroying/modifying/corrupting the source list.

I am shocked and impressed that this happens to work correctly.

Reply:
I can add a `DList_RemoveEntryList` before `DList_InsertTailList(op_list, listEntry)` (line 81) if that makes things more palatable.

Modifying/destroying entries during `DList_ForEach` is explicitly support as per:

`**SRS_DLIST_24_001: [** `DList_ForEach` shall allow deletion of elements during traversal. **]****SRS_DLIST_24_001: [** `DList_ForEach` shall allow deletion of elements during traversal. **]**`